libs/stokesian_dynamics/device_matrix.hpp

```diff
@@ -0,0 +1,658 @@
+    using iterator = pointer;
+
+private:
+    // Storage
+    size_type m_size;
+    T *m_data;
+
+public:
+    device_vector_view(pointer data, size_type size)
+        : m_size(size), m_data(data) {}
+
+    device_vector_view(storage_type &v)
+        : m_size(v.size()), m_data(thrust_wrapper::raw_pointer_cast(v.data())) {}
+
+    DEVICE_FUNC reference operator()(size_type i) noexcept { return m_data[i]; }
+
+    DEVICE_FUNC const_reference operator()(size_type i) const noexcept {
+        return m_data[i];
+    }
+
+    DEVICE_FUNC pointer data() noexcept { return m_data; }
+    DEVICE_FUNC const_pointer data() const noexcept { return m_data; }
+    DEVICE_FUNC size_type size() const noexcept { return m_size; }
+};
+
+/* // trying to move this to thrust_wrapper.hpp
+/// Vector addition 
+template <typename T>
+thrust::device_vector<T> operator+(thrust::device_vector<T> const &x,
+                                   thrust::device_vector<T> const &y) {
+    assert(x.size() == y.size());
+    thrust::device_vector<T> z(x.size());
+    thrust::transform(x.begin(), x.end(), y.begin(), z.begin(),
+                      thrust::plus<T>{});
+    return z;
+}
+
+template <typename T>
+thrust::host_vector<T> operator+(thrust::host_vector<T> const &x,
+                                 thrust::host_vector<T> const &y) {
+    assert(x.size() == y.size());
+    thrust::host_vector<T> z(x.size());
+    thrust::transform(x.begin(), x.end(), y.begin(), z.begin(),
+                      thrust::plus<T>{});
+    return z;
+}
+*/
+
+
+/*
+#ifdef SD_THRUST
+// These cannot be moved to thrust_wrapper, because device_matrix appears
+// Hence, the ifdef
+// Just for debugging anyway?
+template <typename T>
+std::ostream &operator<<(std::ostream &os, 
+                         thrust_wrapper::device_vector<T> const &v) {
+    os << '{';
+    for (std::size_t i = 0; i < v.size(); ++i) {
+        os << std::setw(12) << v[i];
+        if (i < v.size() - 1) {
+            os << ',';
+        }
+    }
+    os << '}';
+    return os;
+}
+
+template <typename T>
+std::ostream &operator<<(std::ostream &os, 
+                         thrust_wrapper::host_vector<T> const &v) {
+    os << '{';
+    for (std::size_t i = 0; i < v.size(); ++i) {
+        os << std::setw(12) << v[i];
+        if (i < v.size() - 1) {
+            os << ',';
+        }
+    }
+    os << '}';
+    return os;
+}
+
+template <typename T, typename Policy>
+std::ostream &operator<<(std::ostream &os, device_matrix<T, Policy> const &m) {
+    os << '{';
+    for (std::size_t i = 0; i < m.rows(); ++i) {
+        if (i > 0) {
+            os << ' ';
+        }
+        os << '{';
+        for (std::size_t j = 0; j < m.cols(); ++j) {
+            os << std::setw(12) << m(i, j);
+            if (j < m.cols() - 1) {
+                os << ',';
+            }
+        }
+        os << '}';
+        if (i < m.rows() - 1) {
+            os << ",\n";
+        }
+    }
+    os << '}';
+    return os;
+}
+#endif
```

Comment:
I won't remove this snippet at the moment, as I suspect @hmenke used it for debugging purposes and it would be unnecessary to rewrite all of it.

Reply:
You can get rid of it.  I'll just throw it in a separate debug header.